feature-BJS2-1610

Kergshi opened the pull request.

Kergshi committed:
на всяк случай 2

Kergshi committed:
ВСЕ РАБОТАЕТ!
Еще и метод удаления проекта добавил
Да-да, удаляет из бд

Kergshi committed:
Работа с проектами

Kergshi committed:
Работа с проектами

Kergshi committed:
Merge remote-tracking branch 'origin/feature-BJS2-1610' into feature-BJS2-1610

Kergshi committed:
Работа с проектами

Kergshi committed:
Увеличено покрытие тестами до 75%
Добавлен Jacoco

InnMish commented on src/main/java/faang/school/projectservice/controller/ProjectController.java:
```diff
@@ -0,0 +1,73 @@
+import faang.school.projectservice.dto.ProjectDto;
+import faang.school.projectservice.dto.ProjectFilterDto;
+import faang.school.projectservice.dto.ProjectUpDateDto;
+import faang.school.projectservice.serviсe.ProjectService;
```
на вот этой строчке мне идея подсвечивает, что в названии пакета service русская буква с
Надо поправить

InnMish commented on src/main/java/faang/school/projectservice/controller/ProjectController.java:
```diff
@@ -0,0 +1,73 @@
+
+    @Operation(parameters = {@Parameter(in = ParameterIn.HEADER, name = "x-user-id", required = true)}, summary = "Create new project")
+    //^ для возможности вручную добавлять id пользователя в сваггере и добавления названия эндпоинтов там же
+    @PostMapping("/projects") // не идемпотентный, всегда создает новую сущность
```
Маппинг тут лучше в единственном числе написать, так как один проект тут создаем

InnMish commented on src/main/java/faang/school/projectservice/controller/ProjectController.java:
```diff
@@ -0,0 +1,73 @@
+    //^ для возможности вручную добавлять id пользователя в сваггере и добавления названия эндпоинтов там же
+    @PostMapping("/projects") // не идемпотентный, всегда создает новую сущность
+    public ProjectDto createProject(@RequestBody ProjectDto projectDto) {
+        projectValidator.validateCreateProject(projectDto);
```
Посмотри, у тебя вот тут проверка в этом методе имеет две проблемы: 
первая в том, что мы решили доставать ownerId из хедера. Сетишь его в дто ты только в сервисе, а проверку вызываешь эту до того, как попадешь в сервис. В этой проверке ты сравниваешь ownerId существующих проектов с ownerId отсюда из самого дто. Но ведь в дто на этом этапе ownerId не заполнен. 

Второй момент: в этом методе ты проверяешь, что такого проекта в бд еще нет по оунеру и по названию. А потом в сервисе на первой прям строчке делаешь то же самое. 

Как будто бы вся вторая часть метода validateCreateProject не нужна вообще именно в этом месте

InnMish commented on src/main/java/faang/school/projectservice/validator/ProjectValidator.java:
```diff
@@ -0,0 +1,76 @@
+    private final ProjectRepository projectRepository;
+
+    public void validateCreateProject(ProjectDto projectDto) {
+        if (projectDto.getName() == null || projectDto.getDescription() == null) {
```
Я думаю, здесь стоит еще добавить проверки на пустые значения 
Типа, если название или описание вот такое: "" 
Или такое " *куча пробелов*"

InnMish commented on src/main/java/faang/school/projectservice/serviсe/ProjectService.java:
```diff
@@ -0,0 +1,91 @@
+
+    @Transactional
+    public ProjectDto createProject(ProjectDto projectDto) {
+        if (projectRepository.existsByOwnerUserIdAndName(projectDto.getOwnerId(), projectDto.getName())) {
```
и вот тут получается, что projectDto.getOwnerId() еще не заполнен значением, потому что только после этой проверки ты заполняешь поле projectDto.setOwnerId() на 36 строчке 

И я сейчас проверила, у меня один и тот же проект добавляется в базу из-за этого. Типа, условие тут не возвращает тру, хотя проект и есть 

Значит, надо сначала засетить owner, а потом уже проверять все.

InnMish commented on src/main/java/faang/school/projectservice/serviсe/ProjectService.java:
```diff
@@ -0,0 +1,91 @@
+    @Transactional
+    public ProjectDto createProject(ProjectDto projectDto) {
+        if (projectRepository.existsByOwnerUserIdAndName(projectDto.getOwnerId(), projectDto.getName())) {
+            throw new DataValidationException("Preject " + projectDto.getName() + " already exist");
```
опечатка в слове Project

InnMish commented on src/main/java/faang/school/projectservice/controller/ProjectController.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    @Operation(parameters = {@Parameter(in = ParameterIn.HEADER, name = "x-user-id", required = true)}, summary = "update project")
+    @PutMapping("/projects/{id}")
```
Тут тоже project в единственном числе - обновляем один конкретный

InnMish commented on src/main/java/faang/school/projectservice/validator/ProjectValidator.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    public void validateUpdateProject(Long id, ProjectUpDateDto projectUpDateDto) {
+        if (id == null || id <= 0 || projectUpDateDto.getDescription() == null && projectUpDateDto.getStatus() == null) {
```
тут Description тоже проверь на непустоту 
Лучше всего через isBlank

InnMish replied:
А еще id null тут, думается мне, не пройдет...) 
Типа, если я пути как PathVariable не укажу id, то мне спринг ругнется на это сразу 
Как будто бы можно вообще убрать эту проверку на null

InnMish commented on src/main/java/faang/school/projectservice/serviсe/ProjectService.java:
```diff
@@ -0,0 +1,91 @@
+    public ProjectDto updateProject(Long id, ProjectUpDateDto projectUpDateDto) {
+        var projectToUpdate = projectRepository.getProjectById(id);
+        projectValidator.validateServiceOwnerOfProject(userContext.getUserId(), projectToUpdate);
+        projectRepository.save(projectMapper.update(projectToUpdate, projectUpDateDto));
```
Мне вот такой подход маппирования кажется усложненным. Ты достаешь из базы здесь сущность, потом ее через маппинг изменяешь, превращая ее (уже красивую и готовую projectToUpdate  с 44) в новую композитную из projectToUpdate  и  projectUpDateDto. 
Это выглядит сложно, потому что внутри маппер перекидывает туда-сюда все вообще поля: вакансии, моменты, этапы и все остальное. И потом мы как будто старый объект выкинем, а новый будем использовать. Операция update подразумевает именно изменение текущего объекта, а не создание нового с таким же Id. 

Сейчас нам известно, что всего два поля можно изменять - описание и статус. Как будто удобнее было бы просто достать из бд вот эту projectToUpdate и засетить туда новые поля. 

Такой подход расширять будет сложнее, если новые поля захотим обновлять, какие-то кроме статуса и описания, но вот прям эффективнее сработает, плюс на той же самой сущности, которую из бд достанешь

InnMish commented on src/main/java/faang/school/projectservice/mapper/ProjectMapper.java:
```diff
@@ -0,0 +1,21 @@
+    @Mapping(target = "description", source = "projectUpDateDto.description")
+    @Mapping(target = "status", source = "projectUpDateDto.status")
+    @Mapping(target = "resources", ignore = true)
+        // если эту^ хреновину не игнорировать, не будет на данном этапе работать
```
Эту хреновину я починила 
Почему-то не было миграции на поля createdBy и updatedBy для resourses  

Добавила в мастер сейчас, подтяни мастер в свою ветку, они появятся тут 
Потом надо заново будет накатить скрипты в бд, чтобы заработало, скорее всего. Ликвибейз может ругаться. 
Быстрее всего будет весь infra заново поднять, но можно и руками в бд положить таблички и заново накатить

InnMish commented on src/main/java/faang/school/projectservice/dto/ProjectUpDateDto.java:
```diff
@@ -0,0 +1,16 @@
+@AllArgsConstructor
+@NoArgsConstructor
+@Data
+public class ProjectUpDateDto {
```
Update - это целое слово такое Обновить  
А у тебя читается как поднять дату) Up Date

InnMish commented on src/main/java/faang/school/projectservice/validator/ProjectValidator.java:
```diff
@@ -0,0 +1,76 @@
+
+    public void validateFilter(ProjectFilterDto projectFilterDto) {
+        if (projectFilterDto.getStatus() == null && projectFilterDto.getName() == null) {
+            throw new DataValidationException("Check for filter");
```
Вот тут текст ошибки надо понятнее сделать 
Типа 
Enter at least one filter

InnMish commented on src/main/java/faang/school/projectservice/controller/ProjectController.java:
```diff
@@ -0,0 +1,73 @@
+    @Operation(parameters = {@Parameter(in = ParameterIn.HEADER, name = "x-user-id", required = true)}, summary = "Get all projects with filters")
+    @PutMapping("/projects/filter")
+    public List<ProjectDto> getAllProjectsWithFilter(@RequestBody ProjectFilterDto projectFilterDto) {
+        projectValidator.validateFilter(projectFilterDto);
```
а вот это точно нужно? У тебя isApplicable на это все равно проверит, и если ни один фильтр не применится, то на выходе просто получится тот же список, что и findAll 
По сути пустые фильтры это не проблема, не такая, по крайней мере, чтобы ошибку кидать. 

Если хочешь, можно, например, если все фильтры null, то можно вызвать метод getAllProjects

InnMish commented on src/main/java/faang/school/projectservice/validator/ProjectValidator.java:
```diff
@@ -0,0 +1,76 @@
+        return projectValidate;
+    }
+
+    public void validateServiceOwnerOfProject(Long userId, Project project) {
```
название тут хочется поменять 
Непонятно, это что за сервис-оунер такой 
можно назвать checkForValidOwner, например

InnMish commented on src/main/java/faang/school/projectservice/validator/ProjectValidator.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    public Project validateServiceGetProject(Long userId, Project project) {
+        Project projectValidate = null;
```
У нас в базе данных хранятся определенные проекты. Мы после фильтрации должны вернуть именно те проекты, которые в базе хранятся, а не какие-то новые, которые мы сами тут создали. 
Чуть ниже пишу, как сделать, чтобы не создавать новый объект

InnMish commented on src/main/java/faang/school/projectservice/validator/ProjectValidator.java:
```diff
@@ -0,0 +1,76 @@
+                        projectValidate = project;
+                        break;
+                    }
+                } // Пишу для очередного желающего сказать, что этот метод никогда не возвращает null.
```
Это для меня?) 
Я даже не поняла, какой метод)

InnMish commented on src/main/java/faang/school/projectservice/validator/ProjectValidator.java:
```diff
@@ -0,0 +1,76 @@
+    public Project validateServiceGetProject(Long userId, Project project) {
+        Project projectValidate = null;
+        List<Team> teams = project.getTeams();
+        if (project.getVisibility() == PRIVATE) {
+            for (Team team : teams) {
+                for (var teamMember : team.getTeamMembers()) {
+                    if (teamMember.getId().equals(userId)) {
+                        projectValidate = project;
+                        break;
+                    }
```
Чтобы удобно определить, возвращать проект или null, можно использовать boolean поле
Проверить, что участник есть - если да, то true, если нет - false  

И в зависимости от этого параметра вернуть проект или null  

 ```
if (project.getVisibility() == PRIVATE) {
            boolean isMember = project.getTeams().stream()
                    .flatMap(team -> team.getTeamMembers().stream())
                    .anyMatch(teamMember -> teamMember.getId().equals(userId));

            if (!isMember) {
                return null;
            }
        }

        return project;
```

Тип того

InnMish commented on src/main/java/faang/school/projectservice/serviсe/ProjectService.java:
```diff
@@ -0,0 +1,91 @@
+                .flatMap(projectFilter -> {
+                    log.info("Applicable filter {}", projectFilter.getClass().getSimpleName());
+                    // Создаем новый стрим перед каждым использованием фильтра, потому что иначе НЕ РАБОТААААЛОООО
+                    Stream<Project> projectsStream = projects.stream()
+                            .map(project -> projectValidator.validateServiceGetProject(userContext.getUserId(), project))
+                            .filter(project -> project != null);
```
.filter(Objects::nonNull);

InnMish replied:
Я бы вот этот участок со Stream<Project> projectsStream = projects.stream().. бла-бла вынесла в отдельный метод для красоты и простоты чтения 

На 64 написать `return projectFilter.applyFilter(projectFilterDto, getFilteredProjects(projects));`

А ниже приватный метод, который стримы вернет в него) (тот код, что с 59 до 61)

InnMish commented on src/main/java/faang/school/projectservice/validator/ProjectValidator.java:
```diff
@@ -0,0 +1,76 @@
+        }
+    }
+
+    public Project validateServiceGetProject(Long userId, Project project) {
```
название метода тоже чуть понятнее давай сделаем
Например, checkIfUserIsMember

InnMish commented on src/main/java/faang/school/projectservice/serviсe/ProjectService.java:
```diff
@@ -0,0 +1,91 @@
+    @Transactional(readOnly = true)
+    public List<ProjectDto> getAllProjects() {
+        List<Project> projects = projectRepository.findAll();
+        return projects.stream()
+                .map(project -> projectValidator.validateServiceGetProject(userContext.getUserId(), project))
+                .map(projectMapper::toDto)
+                .toList();
+    }
```
Смотри, вот тут как раз то же самое действие выполняется, что в методе с фильтрацией - собираешь нужные проекты 

Точно надо в отдельный метод вынести, чтобы переиспользовать потом
Не полностью тут повторение, конечно, но все равно можно как-то красиво придумать

InnMish commented on src/main/java/faang/school/projectservice/controller/ProjectController.java:
```diff
@@ -0,0 +1,73 @@
+    @Operation(parameters = {@Parameter(in = ParameterIn.HEADER, name = "x-user-id", required = true)}, summary = "Delete project by id")
+    @DeleteMapping("/projects/{id}")
+    public void deleteProjectById(@PathVariable Long id) {
+        projectValidator.validateProjectId(id);
```
PathVariable  null не пропустит

InnMish commented on src/main/java/faang/school/projectservice/controller/ProjectController.java:
```diff
@@ -0,0 +1,73 @@
+    @Operation(parameters = {@Parameter(in = ParameterIn.HEADER, name = "x-user-id", required = true)}, summary = "Get project by id")
+    @GetMapping("/projects/{id}") // всегда идемпотентный andPoint
+    public ProjectDto getProjectById(@PathVariable Long id) {
+        projectValidator.validateProjectId(id);
```
PathVariable null не пропустит

InnMish commented on src/main/java/faang/school/projectservice/controller/ProjectController.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    @Operation(parameters = {@Parameter(in = ParameterIn.HEADER, name = "x-user-id", required = true)}, summary = "Get project by id")
+    @GetMapping("/projects/{id}") // всегда идемпотентный andPoint
```
это коммент и не важно, но endpoint пишется)
"конечная точка")

InnMish commented on src/main/java/faang/school/projectservice/serviсe/ProjectService.java:
```diff
@@ -0,0 +1,91 @@
+    }
+
+    @Transactional
+    public void deleteProjectById(Long id) {
```
хе-хе-хе

Я бы не стала именно удалять 
Представь, на проекте работало 500 человек (или прям щас работают), а руководитель такой "ну, народ, до свидания" и удаляет проект))  
 У проекта же статусы есть. Если хотим удалить, можно в cancelled перевести. А дальше нужна обфускация данных - когда в базе мы все данные реальные переносим в отдельную табличку, чтобы сохранить их для истории, а в реальной табличке заменяем их всякой бессмыслицей. Эти две таблицы между собой этой бесмысслицей в виде ключа и связываются)

Kergshi committed:
переделки ,пробема с тестом

Kergshi committed:
Работа над комментами.
2 раза не сохраняются проекты с одним именем для одного пользователя, да и остальное подделал. Метод удаления вообще удалил.

InnMish committed:
Merge branch 'cerberus-master-bc2' into feature-BJS2-1610

# Conflicts:
#	build.gradle.kts
